src/main/kotlin/org/dashevo/dapiclient/DapiClient.kt

```diff
@@ -104,12 +101,116 @@ class DapiClient(var dapiAddressListProvider: DAPIAddressListProvider,
         logger.info("broadcastStateTransition(${stateTransition.toJSON()})")
         val method = BroadcastStateTransitionMethod(stateTransition)
         grpcRequest(method, statusCheck = statusCheck, retryCallback = retryCallback)
     }
 
+    fun broadcastStateTransitionInternal(stateTransition: StateTransition,
+                                         statusCheck: Boolean = false,
+                                         retryCallback: GrpcMethodShouldRetryCallback = DefaultShouldRetryCallback())
+    : BroadcastStateTransitionMethod {
+        logger.info("broadcastStateTransitionInternal(${stateTransition.toJSON()})")
+        val method = BroadcastStateTransitionMethod(stateTransition)
+        grpcRequest(method, statusCheck = statusCheck, retryCallback = retryCallback)
+        return method
+    }
+
+    /**
+     * Wait for state transition result
+     * @param hash
+     * @param prove
+     */
+    fun waitForStateTransitionResult(hash: ByteArray, prove: Boolean): WaitForStateTransitionResult {
+        val method = WaitForStateTransitionResultMethod(hash, prove)
+        logger.info(method.toString())
+        val result = grpcRequest(method) as PlatformOuterClass.WaitForStateTransitionResultResponse
+
+        return if (result.hasError())
+            WaitForStateTransitionResult(StateTransitionBroadcastException(result.error))
+        else WaitForStateTransitionResult(Proof(result.proof))
+    }
+
+    val threadPoolService = Executors.newCachedThreadPool()
+
+    inner class WaitForStateSubmittionCallable(val signedStateTransition: StateTransitionIdentitySigned, val prove: Boolean) :
+        Callable<WaitForStateTransitionResult> {
+        override fun call(): WaitForStateTransitionResult {
+            return try {
+                waitForStateTransitionResult(signedStateTransition.hashOnce(), prove)
+            } catch (e: StatusRuntimeException) {
+                WaitForStateTransitionResult(StateTransitionBroadcastException(e.status.code.value(), e.message?:"", ByteArray(0)))
+            }
+        }
+    }
+
+    fun broadcastStateTransitionAndWait(signedStateTransition: StateTransitionIdentitySigned,
+                                        retriesLeft: Int = USE_DEFAULT_RETRY_COUNT,
+                                        statusCheck: Boolean = false,
+                                        retryCallback: BroadcastShouldRetryCallback = DefaultBroadcastRetryCallback()) {
+
+        val retryAttemptsLeft = if (retriesLeft == USE_DEFAULT_RETRY_COUNT) {
+            retries // set in constructor
+        } else {
+            retriesLeft // if called recursively
+        }
+
+        val futuresList = arrayListOf<Future<WaitForStateTransitionResult>>()
+
+        val futureWithProof = threadPoolService.submit(WaitForStateSubmittionCallable(signedStateTransition, true))
+        futuresList.add(futureWithProof)
```

Comment:
Here we are using a threadpool to call three or more nodes with the API `waitForStateTransitionResult`

Using co-routines here would require the libraries that depend on this code to use coroutines also.  Perhaps that is an action item for a different time.

Reply:
Nice!